### PR TITLE
mech killdozer modkit

### LINKED ISSUE #27615: Feature Request: Killdozer modkit
### What is it?
Killdozer modkit is an item that turns civilian mechs (Ripley, Firefighter Ripley, Odysseus, Clarke) into low cost and stealthy combat mechs
### How does it work?
Inserting a modkit into a mech under maintenance (like batteries or tracking beacons) enables it to use mounted weapons. Visually, the mech sprite is not changed, but if inspected, you can tell the mech has been tampered with (ie. "There are some odd pieces of equipment on this mech" after examining it). However, they also keep the same armor of the civilian models, and are easily taken down by proper combat mechs or weapons.
### Acquiring and using the modkit
Modkits can either be ordered via Syndicate PDA or manufactured after science has done some illegal research. Before using the modkit, it needs a weapons board, similar to how a plasma cutter needs a diamond mining drill (or is it kinetic accelerator) to be completed. Killdozer mechs allow a syndicate agent to make an inconspicuous combat mech that will not raise any eyebrows till its true purpose is revealed. They can also be used as extremely low cost combat mechs when shit hits the fan, like blob or xeno, to quickly get heavy support to the crew

### PR DESCRIPTION
**Pull request by hacker-on-steroids**
## What this does
new item you can attach to exosuits
the killdozer disables requirements on mech modules but uses up a slot - you can put 2 guns on a modified ripley
it also lets you put non-guns onto modified mechs, so you can put a sleeper and a phazon array on a gygax, or 3 banana mortars onto a clarke
closes #27615
image
adds the killdozer to the traitor uplink and r&d(illegal 2)

## Changelog

:cl:
 * rscadd: mecha killdozer kit - bypasses mech module requirements, letting you put guns on working mechs(or tools on combat mechs)
 * rscadd: add killdozer kit to uplink (10tc and comes with 3 random guns)
 * rscadd: add killdozer kit to r&d (illegal 2)

### CODE REVIEW

File: code/game/mecha/equipment/mecha_equipment.dm [hunk 1]
```diff
@@ -97,9 +97,7 @@
 
 /obj/item/mecha_parts/mecha_equipment/proc/can_attach(obj/mecha/M as obj)
 	if(istype(M))
-		if(M.equipment.len<M.max_equip)
```
Comment (west3436) [inline code/game/mecha/equipment/mecha_equipment.dm:100]: why have you done this

File: code/game/mecha/mecha.dm [hunk 2]
```diff
@@ -817,7 +828,7 @@
 	if(istype(W, /obj/item/mecha_parts/mecha_equipment))
 		var/obj/item/mecha_parts/mecha_equipment/E = W
 		spawn()
-			if(E.can_attach(src))
+			if(E.can_attach(src) || is_killdozer()) && get_remaining_equipment_slots())
```
Comment (west3436) [inline code/game/mecha/mecha.dm]: please double check your parentheses this looks wrong

**Comment by west3436** [comment]
This seems pretty low effort - can you at least make a generic KILLDOZER mech sprite that can replace the base mech's sprite?

**Comment by hacker-on-steroids** [comment]
feature request says the mech sprite doesn't change and should only be obvious on close examination
also, making new sprites for every mech/paint combination would be insane(there are 180 in mecha.dmi)

**Comment by west3436** [comment]
> feature request says the mech sprite doesn't change and should only be obvious on close examination also, making new sprites for every mech/paint combination would be insane(there are 180 in mecha.dmi)

No I'm saying make a single sprite instead of a variant for each type (the mech is just covered in concrete)

File: code/game/mecha/equipment/weapons/weapons.dm [hunk 3]
```diff
@@ -637,3 +637,23 @@
...
+		/obj/item/mecha_parts/mecha_equipment/weapon/ballistic/missile_rack/flashbang/inflatable,
+		//leaving the banana mortar in because its funny
+		/obj/item/mecha_parts/mecha_equipment/weapon/honker,
+		/obj/item/mecha_parts/mecha_equipment/weapon/ballistic/missile_rack/mousetrap_mortar,
+		/obj/item/mecha_parts/mecha_equipment/weapon/ballistic/missile_rack/creampie_mortar,
+		///obj/item/mecha_parts/mecha_equipment/weapon/energy/pulse,
+		///obj/item/mecha_parts/mecha_equipment/weapon/ballistic/missile_rack,
+		)
+
+/obj/item/mecha_parts/mecha_equipment/weapon/random_weapon/New()
```
Comment (Kurfursten) [inline code/game/mecha/equipment/weapons/weapons.dm:656]: new must call ..()

**Comment by Kurfursten** [comment]
The main worry I see is using a Clarke as a ghetto phazon. Its very high speed makes it ideal for carrying weapons. As we all know, speed is king - who cares if you have armor and health when you can just evade shots? The clarke moves faster than a human does.

Getting to use the running board on combat mechs is also a pretty big concern.

Stat | Durand | Gygax | Phazon | Clarke |
| -- | ------------- | ------------- | -- | -- |
| Health | 400 | 300 | 200 | 100 |
| Slots | 3  | 3 | 4 | 4
| Speed | 2.5 tps | 3.3 tps | 10 tps | 10tps
| Armors | br .5, fi -.1 , bul .35, las .15, en .1, bom .2 | br .25 fi 0, bul.2 las .3, en .15, bom 0 | br .3, fi .3, bul .3, las .3, en .3, bom .3 | br .2, fi -.2, bul .1, las 0, en 0, bom 0 |

Worse than any one thing this PR might do now, though, I worry that it closes our design space. We can't make a mech module intended for just combat mechs or just working mechs in the future because we know it CAN be used on any mech.

File: code/game/mecha/mecha.dm [hunk 4]
```diff
@@ -817,7 +828,7 @@
 	if(istype(W, /obj/item/mecha_parts/mecha_equipment))
 		var/obj/item/mecha_parts/mecha_equipment/E = W
 		spawn()
-			if(E.can_attach(src))
+			if((E.can_attach(src) || is_killdozer()) && get_remaining_equipment_slots())
```
Comment (Kurfursten) [inline code/game/mecha/mecha.dm:831]: will certainly cause runtimes if you attach a clamp to a combat mech

Comment (hacker-on-steroids) [reply code/game/mecha/mecha.dm:831]: not seeing runtimes but yeah the clamp doesn't work on them

Comment (Kurfursten) [reply code/game/mecha/mecha.dm:831]: lacks cargo capacity

**Comment by hacker-on-steroids** [comment]
image

is this being downboated because it's gay and shit?
should i even bother fixing it?

**Comment by githubuser4141** [comment]
> The main worry I see is using a Clarke as a ghetto phazon. Its very high speed makes it ideal for carrying weapons. As we all know, speed is king - who cares if you have armor and health when you can just evade shots? The clarke moves faster than a human does.
> 
> Getting to use the running board on combat mechs is also a pretty big concern.
> 
> Stat	Durand	Gygax	Phazon	Clarke
> Health	400	300	200	100
> Slots	3	3	4	4
> Speed	2.5 tps	3.3 tps	10 tps	10tps
> Armors	br .5, fi -.1 , bul .35, las .15, en .1, bom .2	br .25 fi 0, bul.2 las .3, en .15, bom 0	br .3, fi .3, bul .3, las .3, en .3, bom .3	br .2, fi -.2, bul .1, las 0, en 0, bom 0
> Worse than any one thing this PR might do now, though, I worry that it closes our design space. We can't make a mech module intended for just combat mechs or just working mechs in the future because we know it CAN be used on any mech.

What's the issue? It has a very low deflect, almost no armor and 100 HP. 100 HP and it's slightly faster then a human but still has to point at a human to hit them. Yes it's dangerous but it's no more dangerous then someone running on meth or phazon with the cap's armor and an egun.

And honestly having generic modules is a step forward. Mechs should be truly multipurpose, why shouldn't you be able to put a sleeper on a Gygax? What we need is to balance the mechs as per their _intended role_, so balance is natural, rather then forcing a role on them which is artificial and hamfisted. 

Why wouldn't you want to use a Durand for mining? Because it's slow as shit, expensive and you can't even use it for combat if it has mining modules so it's pointless, and you look sussy as fuck "mining" with a Durand. So it's POSSIBLE, it's just really unoptimized for the job. This is how all mechs should be balanced, it should be possible to use a Odysseus for combat but you'd never want to because it's super fragile and (can't hold many weapons/armor*).

This is how it's balanced on Baystation - BTW. Actually, it's balanced by two points - interchangeable mech software boards (utility, medical, engi and weapons) - but since a mech can hold up to 2 boards each it's still possible to have a MediGax, or a Duripely, or a Phlarke. The other point being that combat mech components are by their nature heavier, more expensive and require more power to operate then their lighter counterparts. So for rescue/utility purposes you'd naturally gravitate towards the weaker components. You can ofc ignore this at the cost of everyone being annoyed and weirded out there's a durand moving crates, and the fact that even if you don't get hosed with ions it's going to take more actual time. Like to move one box it's going to take 5 IRL minutes instead of like 1 IRL minute, which sucks. Like you're legit wasting your irl life trying to powergame (w-what if the cargo technician shoots me) by using a Durand to haul crates to departments.

Now IMHO for a 'generalized mech gear' thingy to properly work, without exploits/tricks we'd need mech in hand equipment sprites. So you can't BOOO surprise people by having a clarke with 2 machineguns. Equipment sprites at the moment aren't a serious balance issue because you can't do that, so you see a Clarke and assume it's harmless, you see a Gygax and assume it's armed and probably trying to kill you. We'd probably want to add other balances like, weight and maybe "hardpoint size" to force certain mechs to be innately less efficient at equipping certain gear, as in it'd slow them down, occupy more then 1 slot, etc etc.

**Comment by githubuser4141** [comment]
Also why is it called a killdozer, the killdozer irl was an armour package + gunports attached to a bulldozer, yeah I get the bulldozer-mech thinking but this doesn't add armour at all?

**Comment by west3436** [comment]
> Also why is it called a killdozer, the killdozer irl was an armour package + gunports attached to a bulldozer, yeah I get the bulldozer-mech thinking but this doesn't add armour at all?

Yeah that's why I told him to make a generic concrete armor sprite that replaces the mech. There doesn't even need to be a unique sprite for each mech just make it a concrete box lmao

**Comment by Blithering** [comment]
I'd personally say to limit it to Ripleys and give it a visible kilkdozered up sprite overlay, but sounds fun, could be worth test merging at least.

**Comment by hacker-on-steroids** [comment]
drafting until anonymous spoonfeeds me a sprite
making it apply only to ripleys should also stop the clamp nervegas

**Comment by Kurfursten** [comment]
I feel that restricting it to Ripley will also address my concerns

**Comment by Blithering** [comment]
I can add it to my to be sprited list since I'll be touching mechs soon™, if anyone else wants to make one faster please do.

**Comment by hacker-on-steroids** [comment]
anonymous delivered
image
now i just need to get around to doing this

**Comment by west3436** [comment]
@hacker-on-steroids ever gonna come back to this? The new sprite looks great

**Comment by hacker-on-steroids** [comment]
playing minecraft rn gimme a bit

**Comment by west3436** [comment]
@hacker-on-steroids hello i hope you've enjoyed the first 4000 hours of GregTech (early game) but do you plan to finish this??

**Comment by Blithering** [comment]
I hope he's having fun on Minecraft.

**Comment by hacker-on-steroids** [comment]
i'm back from gulag, might finish this soon

**Comment by wotisjanitor** [comment]
Test merge, IMO. This has potential to be extremely annoying but let's find out

**Comment by hacker-on-steroids** [comment]
we'll see which one the people want #36860 #36861 #36862

**Pull request closed without merging**